official/utils/testing/golden.py

```diff
@@ -0,0 +1,279 @@
+just calls the op definition function with test=False for all relevant tests.
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+
+import argparse
+import hashlib
+import json
+import os
+import shutil
+import warnings
+
+
+import numpy as np
+import tensorflow as tf
+from tensorflow.python import pywrap_tensorflow
+
+
+class BaseTest(tf.test.TestCase):
+  """TestCase subclass for performing golden tests.
+  """
+
+  def regenerate(self):
+    """Subclasses should override this class to generate new reference data.
+    """
+    raise NotImplementedError
+
+  @property
+  def file(self):
+    """Subclasses should override to "return __file__".
+    """
+    raise NotImplementedError
+
+  @property
+  def data_root(self):
+    """Use the subclass directory rather than the parent directory.
+
+    Returns:
+      The path prefix for reference data.
+    """
+    return os.path.join(os.path.split(
+        os.path.abspath(self.file))[0], "test_data")
+
+  @property
+  def ckpt_prefix(self):
+    return "model.ckpt"
+
+  @staticmethod
+  def name_to_seed(name):
+    seed = hashlib.md5(name.encode("utf-8")).hexdigest()
+    return int(seed, 16) % (2**32 - 1)
+
+  @staticmethod
+  def common_tensor_properties(input_array):
+    """Convenience function for matrix testing.
+
+    Args:
+      input_array: Tensor (numpy array), from which key values are extracted.
+
+    Returns:
+      A list of key values.
+    """
+    output = list(input_array.shape)
+    flat_array = input_array.flatten()
+    output.extend([float(i) for i in
+                   [flat_array[0], flat_array[-1], np.sum(flat_array)]])
+    return output
+
+  def default_correctness_function(self, *args):
+    """Returns a vector with the concatenation of common properties.
+
+    Args:
+      *args: A list of numpy arrays corresponding to tensors which have been
+        evaluated.
+    """
+    output = []
+    for arg in args:
+      output.extend(self.common_tensor_properties(arg))
+    return output
+
+  def _construct_test_case(self, name, graph, ops_to_eval,
+                           correctness_function):
+    """Save a gold standard.
+
+    Args:
+      name: String defining the run. This will be used to define folder names
+        and will be used for random seed construction.
+      graph: The graph in which the test is conducted.
+      ops_to_eval: Ops which the user wishes to be evaluated under a controlled
+        session.
+      correctness_function: This function accepts the evaluated results of
+        ops_to_eval, and returns a list of values. This list must be JSON
+        serializable; in particular it is up to the user to convert numpy
+        dtypes into builtin dtypes.
+    """
+    data_dir = os.path.join(self.data_root, name)
+
+    # Make sure there is a clean space for results.
+    if os.path.exists(data_dir):
+      shutil.rmtree(data_dir)
+    os.mkdir(data_dir)
```

Comment:
Probably worth to be os.makedirs in case the parent dir is not created.

Reply:
Yes. Fixed.